llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -329,6 +334,51 @@ static void emitDirectiveRelocJalr(const MachineInstr &MI,
   }
 }
 
+// NOTE: This is being used to control optimization of BALC in the linker.
+// TODO: Introduce a new relocation that will serve for this purpose only.
+static bool emitDirectiveRelocBalcStubs(const MachineInstr &MI,
+                                        MCContext &OutContext,
+                                        TargetMachine &TM,
+                                        MCStreamer &OutStreamer) {
+  assert(EmitNanoMipsBalcStubsReloc);
+  if (MI.getMF() && MI.getMF()->getFunction().hasOptSize()) {
+    MCSymbol *OffsetLabel = OutContext.createTempSymbol();
+    const MCExpr *OffsetExpr = MCSymbolRefExpr::create(OffsetLabel, OutContext);
+    // This symbol will be used to indicate that we do want balc stubs
+    // optimization. The '\2' ensures that no other symbol will get the same
+    // name as this.
+    MCSymbol *DummySymbol =
+        OutContext.getOrCreateSymbol("__reloc_balc_stub_\\2");
+    DummySymbol->setExternal(true);
+    OutStreamer.emitSymbolAttribute(DummySymbol, MCSA_Global);
+
+    if (!MI.getOperand(0).isGlobal())
```

Comment:
This will pass over compiler builtins, where no C declaration exists. Consider memcpy/memset and libgcc primitives.

Reply:
This will be removed.